### PR TITLE
Add ImmutableDefinitionChecker

### PR DESCRIPTION
**Pull request by j3parker**

### CODE REVIEW

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+	internal sealed class ImmutableDefinitionChecker {
+		private readonly SemanticModel m_model;
+		private readonly Action<Diagnostic> m_diagnosticSink;
+		private readonly ImmutabilityContext m_context;
+
+		public ImmutableDefinitionChecker(
+			SemanticModel model,
+			Action<Diagnostic> diagnosticSink,
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:15]: Why this diagnosticSink stuff?

Having nested generators made the stuff uglier.

Callers would just pass `m_ctx.ReportDiagnostic` here.

File: src/D2L.CodeStyle.Analyzers/Diagnostics.cs [hunk 2]
```diff
@@ -597,5 +597,14 @@ public static class Diagnostics {
 			defaultSeverity: DiagnosticSeverity.Error,
 			isEnabledByDefault: true
 		);
+
+		public static readonly DiagnosticDescriptor EventMemberMutable = new DiagnosticDescriptor(
+			id: "D2L0070",
+			title: "Event members are not allowed in immtuable types.",
+			messageFormat: "Event members hold registered event handlers. We consider this to be a source of mutability.",
+			category: "Immtubility",
```
Comment (mthjones) [inline src/D2L.CodeStyle.Analyzers/Diagnostics.cs:605]:
```suggestion
			category: "Immutability",
```

(nit: maybe a shared `const` would be good for this?)

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+		) {
+			m_model = model;
+			m_diagnosticSink = diagnosticSink;
+			m_context = context;
+		}
+
+		public bool CheckStruct( StructDeclarationSyntax decl ) => CheckDeclaration( decl );
+		public bool CheckClass( ClassDeclarationSyntax decl ) => CheckDeclaration( decl );
+
+		// This method is private to ensure callers only pass structs & classes
```
Comment (mthjones) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:26]: 👍

**Review by mthjones — CHANGES_REQUESTED** [review]

File: src/D2L.CodeStyle.Analyzers/Diagnostics.cs (hunk 2 above)
Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/Diagnostics.cs]: Probably a good idea -- @omsmith and I were talking about having DI pick the analyzers to run based on this category (by looking at their `SupportedDiagnostics`)

**Review by omsmith — APPROVED** [review]

**Pull request merged**